### PR TITLE
Blankenbach Benchmark

### PR DESCRIPTION
**Pull request by tjhei**
This adds the Blankenbach cases 1a, 1b, 1c, 2a, 2b

Questions:
I added a plugin that writes the current temperature in an ascii format that can be used to start the solution closer to the correct state. It is still in a messy state, because there are some issues:
- even when storing the final solution, I need 1000+ timesteps when starting from it (even if it is done on a finer mesh, see readme.txt). I am not sure what the problem is.
- I have an experimental version (commented in the code) that outputs the solution in the vertices instead of the cell midpoints (turns out that gives worse results and is less elegant)
- It only works in serial and without adaptive refinement
- floating point sort/unique are not robust, but I am not sure how to fix that

Other things:
- Right now I created a large number of .prm files. Is that acceptable?
- The average temperature doesn't make sense, I think blankenbach contains the surface temperature?
- Do we want other statistics?
- Should we implement a fixed point iteration instead? Solving thousands of time steps is not a great solution for this kind of problems.

### CODE REVIEW

**Comment by gassmoeller** [comment]
Hi Timo, very nice and a lot of work :+1:. 

- I am not very concerned about the number of files, each is for a different case or resolution. I usually resolve this with a shell script that does something like `cat file$part1.prm file$part2.prm ... | aspect --`, but this does not fit into the makefile approach you used here, and the makefile is more useful to avoid running all of the benchmarks again if only one needs to be remade. Maybe the files can be a bit shortened/cleaned of unnecessary comments, but otherwise I think it is fine.
- What are you measuring to know when to stop the computation? When looking in the Blankenbach paper they do not seem to mention any criterion, so I guess it is somewhat arbitrary? Would a looser criterion reduce the timesteps a lot?
- For now, I would move the temperature output postprocessor into the benchmark folder as a shared library. There is simply too much work to be done to make this of general use, and it is not related to the PR. It is a great starting point for further work in that direction though.
- The average temperature you mentioned confused me as well, until I read the paper again. T_e is not the average temperature, but it is the minimum temperature on the line at x=0.5 close to the bottom boundary (no clue who had that idea). z_e is the corresponding z-coordinate of the temperature minimum on that line. I did not try to automate the computation, but a quick check with 'Plot over line' in Paraview showed that we approximately match the position (z approx 0.22) and temperature (T approx 0.4) for case 1a and refinement level 4. Do you think we need to show convergence for these properties as well?
- Other statistics: q1 and q2 might be interesting, but would need to be computed within the code (temperature gradients at the corners). topography and geoid are not that much important, although the geoid might actually be a good benchmark when we finally continue our geoid effort.
- Iterations: The paper states that some participants used fixed-point iterations while others used timesteps. I do not think it is worth the effort to redo everything. It might be faster, but is it worth the effort?

I can give this PR a proper review tomorrow.

**Comment by gassmoeller** [comment]
Concerning the number of files: Alternatively we could only provide the base for each of the 5 cases, and provide a shell script that then creates all of the files necessary to run all cases. The makefile could then remove all of the temporary files when it creates the output. Not sure if its worth the effort, but it would bring down the number of files.

**Comment by tjhei** [comment]
Good comments, thanks!

> What are you measuring to know when to stop the computation? When looking in the Blankenbach paper they do not seem to mention any criterion, so I guess it is somewhat arbitrary? Would a looser criterion reduce the timesteps a lot?

I am using the steady state detection termination criterion. It is set to RMS relative difference of 1e-5. If you increase that, you need fewer time steps. But we want the RMS to be exact up to 4-5 digits on the finest mesh...
Especially the case 2b is quite terrible (this is with a "good" starting value):

blankenbach-case2b

> For now, I would move the temperature output postprocessor into the benchmark folder as a shared library. There is simply too much work to be done to make this of general use, and it is not related to the PR. It is a great starting point for further work in that direction though.

You mean as a plugin in benchmark/blankenbach/? Good idea.

> The average temperature you mentioned confused me as well, until I read the paper again. T_e is not the average temperature, but it is the minimum temperature on the line at x=0.5 close to the bottom boundary (no clue who had that idea). z_e is the corresponding z-coordinate of the temperature minimum on that line. I did not try to automate the computation, but a quick check with 'Plot over line' in Paraview showed that we approximately match the position (z approx 0.22) and temperature (T approx 0.4) for case 1a and refinement level 4. Do you think we need to show convergence for these properties as well?

Oh, oops. Maybe we should just skip it.

> Other statistics: q1 and q2 might be interesting, but would need to be computed within the code (temperature gradients at the corners). topography and geoid are not that much important, although the geoid might actually be a good benchmark when we finally continue our geoid effort.

Yes, it is probably more useful from that standpoint (can we compute these statistics).

> Iterations: The paper states that some participants used fixed-point iterations while others used timesteps. I do not think it is worth the effort to redo everything. It might be faster, but is it worth the effort?

To be honest, I am not really thrilled to spend a lot of time on this benchmark. On the other hand, it would be great to have it "done" and converge to more accurate answers than what we have right now.

**Comment by tjhei** [comment]
> Concerning the number of files: Alternatively we could only provide the base for each of the 5 cases, and provide a shell script that then creates all of the files necessary to run all cases. The makefile could then remove all of the temporary files when it creates the output. Not sure if its worth the effort, but it would bring down the number of files.

You have a point here. Let's see.

**Comment by tjhei** [comment]
updated.

**Comment by tjhei** [comment]
One thing that we should try is exporting the temperature in the support points and see if that allows us to resume exactly at the steady state. If not, something else is going wrong with how we set up the computation (maybe the first step with dt=0?).

**Comment by tjhei** [comment]
> Feel free to try the support point export / restart approach,

I tried but but even though the timestep 0 and the final timestep (in this test 137) are identical, I am getting a small movement in T,u, and p in timestep 1 that takes a while to decay. @bangerth I think this shouldn't happen based on the way we solve the first timestep, right? Any idea what is going on? Maybe because the timestep 1 uses backward Euler? Maybe a bug? Or is this just a common thing to have if the solution is not 100% identical?

**Comment by tjhei** [comment]
Interestingly, the left/right heat flux is 10% different in timestep 1 compared to timestep 0 and 137 (the latter two have identical statistics).

**Comment by tjhei** [comment]
The reason seems to be that the entropy viscosity is computed differently in step 1 and is a lot larger. ;-(

**Comment by gassmoeller** [comment]
That makes sense, although it is surprising that this perturbation is even larger than the export/import perturbation of the temperature field. Looks like we need to investigate one more time the initialization behavior of the entropy-viscosity calculation. I feel like this is a larger issue though that should be handled outside of this PR. Even if we find a better way it is likely to change every test.

**Comment by tjhei** [comment]
agreed.

**Comment by gassmoeller** [comment]
Ok, then feel free to merge when you are happy with the current state. I am just waiting to pull the trigger because of the [WIP] in the title.

**Comment by tjhei** [comment]
I was hoping @bangerth would chime in with an idea. But let's merge and open an issue.

**Comment by bangerth** [comment]
I can try to look at this over the weekend, but you probably noticed my absence from much discussion this week -- my calendar is looking pretty terrible and everything is backed up :-(

**Comment by tjhei** [comment]
No hurry for that, let's look at #1389 at some point and merge this PR as is.

**Comment by gassmoeller** [comment]
Something went wrong during your squash. Please fix and merge yourself.

**Comment by tjhei** [comment]
> Something went wrong during your squash. Please fix and merge yourself.

No, the problem is that "astyle-tester" doesn't indent or warn about files in tests/ and benchmark/, I noticed that two days ago and enabled "aspect-astyle-hd" again, that does that correctly (by calling doc/indent). Anyways, fixed.

File: benchmark/blankenbach/Makefile [hunk 1]
```diff
@@ -0,0 +1,38 @@
+.PHONY: all
```
Comment (bangerth) [inline benchmark/blankenbach/Makefile]: This whole Makefile is a bit obscure. A couple of lines at the top providing an overview of what it does would be helpful.

File: benchmark/blankenbach/base_case1b.prm [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  set Model name = ascii data
+  subsection Ascii data model
+    set Data directory = 
+    set Data file name = initial_temperature_case1b.txt
+  end
+end
+
+
+subsection Material model
+  subsection Nondimensional model
```
Comment (bangerth) [inline benchmark/blankenbach/base_case1b.prm]: maybe just to make clear that this is the model that's actually used, repeat the `set Model = nondimensional` line here?

File: benchmark/blankenbach/base_case1c.prm [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  set Model name = ascii data
+  subsection Ascii data model
+    set Data directory = 
+    set Data file name = initial_temperature_case1c.txt
+  end
+end
+
+
+subsection Material model
+  subsection Nondimensional model
```
Comment (bangerth) [inline benchmark/blankenbach/base_case1c.prm]: same here

File: benchmark/blankenbach/plugin/code.cc [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+      return std::make_pair (std::string ("Writing TemperatureAsciiOut:"),
+                             filename);
+    }
+
+
+    template <int dim>
+    void
+    TemperatureAsciiOut<dim>::declare_parameters (ParameterHandler &/*prm*/)
+    {
+    }
```
Comment (bangerth) [inline benchmark/blankenbach/plugin/code.cc]: You know that you don't actually have to implement this and the next function if their bodies are empty, right? There is a default implementation in the base class that already provides the empty body.

File: benchmark/blankenbach/plugin/code.cc [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+// explicit instantiations
+namespace aspect
+{
+  namespace Postprocess
+  {
+    ASPECT_REGISTER_POSTPROCESSOR(TemperatureAsciiOut,
+                                  "temperature ascii out",
+                                  "A postprocessor that generates ascii data "
+                                  "output of the temperature field to be "
+                                  "used as initial condition.")
```
Comment (bangerth) [inline benchmark/blankenbach/plugin/code.cc]: ...that can be read in in a subsequent computation

File: benchmark/blankenbach/readme.txt [hunk 6]
```diff
@@ -0,0 +1,50 @@
+Blankenback benchmark, see table 9 for reference values
+
+Output:
+# Nu           Vrms           name:
+4.78661864e+00 4.34590432e+01 case1a_ref4.stat
```
Comment (bangerth) [inline benchmark/blankenbach/readme.txt]: The comment at the top doesn't give much away -- are these *our* results, or are they the ones from the paper?

File: benchmark/blankenbach/readme.txt [hunk 7]
```diff
@@ -0,0 +1,50 @@
+Blankenback benchmark, see table 9 for reference values
```
Comment (bangerth) [inline benchmark/blankenbach/readme.txt]: table 9 *in the Blankenbach paper*?

**Comment by bangerth** [comment]
Did you intend to add a section to the manual about these benchmarks?

**Comment by tjhei** [comment]
> Did you intend to add a section to the manual about these benchmarks?

I was not planning to do that because there is nothing really interesting to learn/see from it and it wouldn't be the first benchmark to not be described in the manual. If you want I can put the table and a picture from the solution into the manual though.

**Comment by bangerth** [comment]
Yes, I think a short section would be nice just to make sure that the existence of these benchmarks are linked to and visible to the readers of the manual. The section need not be long.

**Pull request merged**